File: frontend/types/invoice.ts
export type InvoiceStatus = "draft" | "sent" | "paid";
export type InvoiceItemType = "rent" | "electricity" | "water" | "surcharge" | "shared_elec";

export const INVOICE_STATUS_LABELS: Record<InvoiceStatus, string> = {
  draft: "Nháp",
  sent: "Đã gửi",
  paid: "Đã thanh toán",
};

export const INVOICE_STATUS_COLORS: Record<InvoiceStatus, { bg: string; fg: string; dot: string }> = {
  draft:  { bg: "var(--slate-100)", fg: "var(--slate-600)",  dot: "var(--slate-400)" },
  sent:   { bg: "var(--blue-50)",   fg: "var(--blue-700)",   dot: "var(--blue-500)" },
  paid:   { bg: "var(--green-50)",  fg: "var(--green-700)",  dot: "var(--green-500)" },
};

export const INVOICE_ITEM_LABELS: Record<InvoiceItemType, string> = {
  rent:        "Tiền thuê",
  electricity: "Tiền điện",
  water:       "Tiền nước",
  surcharge:   "Phụ phí",
  shared_elec: "Điện chung",
};

export type InvoiceItem = {
  id: number;
  item_type: InvoiceItemType;
  name: string;
  unit_price: string;
  quantity: string;
  amount: string;
};

export type Invoice = {
  id: number;
  contract_id: number;
  period: string;
  total: string;
  status: InvoiceStatus;
  public_token: string;
  payment_reported_at: string | null;
  items: InvoiceItem[];
};

export type InvoiceListItem = Invoice & {
  room_id: number;
  room_number: string;
  property_name: string;
  tenant_name: string;
};

File: backend/app/models/invoice.py
from sqlmodel import SQLModel, Field
from decimal import Decimal
from datetime import datetime
from enum import Enum
import uuid


class InvoiceStatus(str, Enum):
    draft = "draft"
    sent = "sent"
    paid = "paid"


class InvoiceItemType(str, Enum):
    rent = "rent"
    electricity = "electricity"
    water = "water"
    surcharge = "surcharge"
    shared_elec = "shared_elec"


class Invoice(SQLModel, table=True):
    __tablename__ = "invoice"

    id: int | None = Field(default=None, primary_key=True)
    contract_id: int = Field(foreign_key="contract.id", index=True)
    period: str = Field(max_length=7)              # "YYYY-MM"
    total: Decimal = Field(decimal_places=0, max_digits=14)
    status: InvoiceStatus = Field(default=InvoiceStatus.draft)
    public_token: str = Field(default_factory=lambda: str(uuid.uuid4()), index=True)
    payment_reported_at: datetime | None = Field(default=None)


class InvoiceItem(SQLModel, table=True):
    __tablename__ = "invoice_item"

    id: int | None = Field(default=None, primary_key=True)
    invoice_id: int = Field(foreign_key="invoice.id", index=True)
    item_type: InvoiceItemType
    name: str
    unit_price: Decimal = Field(decimal_places=2, max_digits=14)
    quantity: Decimal = Field(default=Decimal("1"), decimal_places=2, max_digits=10)
    amount: Decimal = Field(decimal_places=0, max_digits=14)
